python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+
+
+class DoubleLiteral(Literal[float]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DoubleType)
+    def _(self, type_var: DoubleType) -> "DoubleLiteral":
+        return self
+
+    @to.register(FloatType)
+    def _(self, type_var: FloatType) -> Union[AboveMax, BelowMin, FloatLiteral]:
+        if FloatType.max < self.value:
+            return AboveMax()
+        elif FloatType.min > self.value:
+            return BelowMin()
+        return FloatLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> "DecimalLiteral":
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DateLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return self
+
+
+class TimeLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return self
+
+
+class TimestampLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> "TimestampLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.fromtimestamp(self.value / 1000000) - EPOCH).days)
+
+
+class DecimalLiteral(Literal[Decimal]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> Optional["DecimalLiteral"]:
+        if type_var.scale == abs(self.value.as_tuple().exponent):
+            return self
+        return None
+
+
+class StringLiteral(Literal[str]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(StringType)
+    def _(self, type_var: StringType) -> "StringLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return TimeLiteral(
+            int(
+                (
+                    datetime.datetime.strptime((EPOCH.strftime("%Y-%m-%d ") + self.value), "%Y-%m-%d %H:%M:%S.%f") - EPOCH
+                ).total_seconds()
+                * 1000000
+            )
+        )
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> TimestampLiteral:
```

Comment:
I think there's an easier way to do this:

```python
from datetime import datetime
import re

EPOCH = datetime.fromisoformat("1970-01-01T00:00:00.000000")
ISO_TIMESTAMP = re.compile("\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d(.\d{1,6})?")
...
    if ISO_TIMESTAMP.fullmatch(self.value):
        try:
            delta = (datetime.fromisoformat(self.value) - EPOCH)
            return (delta.days * 86400 + delta.seconds) * 1_000_000 + delta.microseconds
        except TypeError:
            return None
    return None
```

Unfortunately, `fromisoformat` is a bit too permissive, as is `strptime`. The problem with `fromisoformat` is that it accepts partial ISO timestamps, like `"2022-03-28"` and doesn't validate that all fields are there. The problem with `strptime` is that it accepts strings that are missing digits, like `2022-3-28 1:2:3`. That's why I think we should use `fromisoformat`, but strictly validate the incoming timestamp string using a regular expression.

This also uses logic similar to `total_seconds` but doesn't needlessly convert to float.